src/pure_pursuit_node.cpp

```diff
@@ -2,37 +2,144 @@
 {
-  path_subscription_ = this->create_subscription<nav_msgs::msg::Path>(
-    "/path", 10,
+  path_subscription = this->create_subscription<nav_msgs::msg::Path>(
+    "/plan", 10,
     std::bind(&PurePursuitNode::path_callback, this, std::placeholders::_1));
-   
-  velocity_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>(
-    "/cmd_vel", 10);
+
+  velocity_publisher = this->create_publisher<geometry_msgs::msg::Twist>(
+    "/vel", 1);
+
+  carrot_publisher = this->create_publisher<geometry_msgs::msg::PointStamped>(
+    "/lookahead_point", 1);
+
+  this->declare_parameter<float>("lookahead_distance", 0.1f);
+  this->declare_parameter<float>("desired_linear_velocity", 1.0);
+  this->declare_parameter<float>("desired_angular_velocity", 1.0);
+  this->declare_parameter<std::string>("robot_frame", "base_footprint");
+  this->declare_parameter<std::string>("map_frame", "map");
+  this->declare_parameter<bool>("m_path_is_initialized", false);
+  this->declare_parameter<float>("tf_timeout", 0.03);
+
+  tf_buffer = std::make_unique<tf2_ros::Buffer>(this->get_clock());
+  transform_listener = std::make_shared<tf2_ros::TransformListener>(*tf_buffer);
+
+  param_update_timer = this->create_wall_timer(
+    1000ms, std::bind(&PurePursuitNode::param_update, this)
+  );
+}
+
+void PurePursuitNode::param_update()
+{
+  this->get_parameter("lookahead_distance", lookahead_distance);
+  this->get_parameter("desired_linear_velocity", desired_linear_velocity);
+  this->get_parameter("desired_angular_velocity", desired_angular_velocity);
+  this->get_parameter("tf_timeout", tf_timeout);
 }
 
 void PurePursuitNode::path_callback(const nav_msgs::msg::Path::SharedPtr msg)
 {
-  last_path_state_ = msg;
-	std::cout << " Here lmao " << std::endl; // TODO(dcutting133): :thinking:
+  m_path = to_path(msg);
+  m_tracker.reset_path(m_path);
+
+  if(!m_path_is_initialized) 
+  {
+    RCLCPP_INFO(this->get_logger(), "Got Path:)");
+    m_path_is_initialized = true;
+  }
+  compute_velocity();
+}
+
+void PurePursuitNode::compute_velocity()
+{
+  if(m_path_is_initialized)
+  {
+    geometry_msgs::msg::Twist velocity;
+    auto [lookahead, heading_to_point, heading_error] = m_tracker.get_target_state(get_pose());
+    velocity.linear.x = lookahead.x;
+    velocity.linear.y = lookahead.y;
+    velocity.linear.z = lookahead.z;
+
+    double angular_vel = constrain(heading_to_point*heading_error, -3.14, 3.14); // TODO parameter these bad bois
+    velocity.angular.z = angular_vel;
+    velocity_publisher->publish(velocity);
+
+    geometry_msgs::msg::PointStamped lookahead_point;
+    lookahead_point.header.stamp = this->get_clock()->now();
+    lookahead_point.header.frame_id = "map";
+    lookahead_point.point.x = lookahead.x;
+    lookahead_point.point.y = lookahead.y;
+    lookahead_point.point.z = 0.3; // TODO parameter this boy
+    carrot_publisher->publish(lookahead_point);
+
+  }
+}
+
+double PurePursuitNode::constrain(float x, float x_min, float x_max)
```

Comment:
I'm not seeing where this is used in the node. Does it belong in the actual PP library part?

Reply:
The constrain function? i actually replaced it with std::clamp. the PP library doesn't use it but the ros2 wrapper does. it is to ensure the angular velocity is within a certain threshold

Reply:
could you remove this code if youre not using it?

Reply:
yes. It will be in my next commit lmao